### ISSUE TITLE
ClassNotFoundException in rdf4j-server using Docker image

### ISSUE THREAD

**Issue by hugovanversendaal** [eclipse-rdf4j/rdf4j#4742]
### Current Behavior

Requests to /rdf4j-server result in an internal server error (500) when using the Docker image eclipse/rdf4j-workbench. The root cause is "java.lang.ClassNotFoundException: javax.servlet.jsp.jstl.core.Config". Below is the server log:

<details>
  <summary>localhost.2023-08-09.log</summary>
<pre><code>
09-Aug-2023 09:09:51.634 INFO [localhost-startStop-1] org.apache.catalina.core.ApplicationContext.log No Spring WebApplicationInitializer types detected on classpath
09-Aug-2023 09:09:51.746 INFO [localhost-startStop-1] org.apache.catalina.core.ApplicationContext.log org.tuckey.web.filters.urlrewrite.UrlRewriteFilter INFO: status display disabled
09-Aug-2023 09:09:51.766 INFO [localhost-startStop-1] org.apache.catalina.core.ApplicationContext.log org.tuckey.web.filters.urlrewrite.UrlRewriteFilter INFO: loaded (conf ok)
09-Aug-2023 09:09:51.867 INFO [localhost-startStop-1] org.apache.catalina.core.ApplicationContext.log Initializing Spring DispatcherServlet 'rdf4j-http-server'
09-Aug-2023 09:10:03.082 SEVERE [http-nio-8080-exec-3] org.apache.catalina.core.StandardWrapperValve.invoke Servlet.service() for servlet [rdf4j-http-server] in context with path [/rdf4j-server] threw exception [Handler processing failed; nested exception is java.lang.NoClassDefFoundError: javax/servlet/jsp/jstl/core/Config] with root cause
	java.lang.ClassNotFoundException: javax.servlet.jsp.jstl.core.Config
		at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1420)
		at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1228)
		at org.springframework.web.servlet.support.JstlUtils.exposeLocalizationContext(JstlUtils.java:103)
		at org.springframework.web.servlet.view.JstlView.exposeHelpers(JstlView.java:137)
		at org.springframework.web.servlet.view.InternalResourceView.renderMergedOutputModel(InternalResourceView.java:145)
		at org.springframework.web.servlet.view.AbstractView.render(AbstractView.java:316)
		at org.springframework.web.servlet.DispatcherServlet.render(DispatcherServlet.java:1405)
		at org.springframework.web.servlet.DispatcherServlet.processDispatchResult(DispatcherServlet.java:1149)
		at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1088)
		at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:964)
		at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
		at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)
		at javax.servlet.http.HttpServlet.service(HttpServlet.java:489)
		at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
		at javax.servlet.http.HttpServlet.service(HttpServlet.java:583)
		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:212)
		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:156)
		at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51)
		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:181)
		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:156)
		at org.apache.catalina.filters.SetCharacterEncodingFilter.doFilter(SetCharacterEncodingFilter.java:115)
		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:181)
		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:156)
		at org.tuckey.web.filters.urlrewrite.RuleChain.handleRewrite(RuleChain.java:176)
		at org.tuckey.web.filters.urlrewrite.RuleChain.doRules(RuleChain.java:145)
		at org.tuckey.web.filters.urlrewrite.UrlRewriter.processRequest(UrlRewriter.java:92)
		at org.tuckey.web.filters.urlrewrite.UrlRewriteFilter.doFilter(UrlRewriteFilter.java:389)
		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:181)
		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:156)
		at com.github.ziplet.filter.compression.CompressingFilter.doFilter(CompressingFilter.java:263)
		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:181)
		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:156)
		at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:167)
		at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90)
		at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:483)
		at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:130)
		at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93)
		at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:682)
		at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
		at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
		at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:617)
		at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63)
		at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:932)
		at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1694)
		at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)
		at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
		at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
		at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
		at java.base/java.lang.Thread.run(Unknown Source)
</code></pre>
</details>

### Expected Behavior

Requests to /rdf4j-server do not result in an error.

### Steps To Reproduce

1. `docker pull eclipse/rdf4j-workbench:4.3.5`
2. `docker run -d -p 8080:8080 -e JAVA_OPTS="-Xms1g -Xmx4g" -v data:/var/rdf4j -v logs:/usr/local/tomcat/logs eclipse/rdf4j-workbench:4.3.5`
3. Open a webbrowser and go to http://localhost:8080/rdf4j-server or run `curl -L http://localhost:8080/rdf4j-server`

### Version

4.3.5

### Are you interested in contributing a solution yourself?

None

**Comment by abrokenjester** [comment]
I have noticed this same problem affects RDF4J Server when deployed directly in Tomcat 9. The server starts, but its rendered pages do not work.

**Comment by abrokenjester** [comment]
Possibly related to/caused by fixes for https://github.com/eclipse-rdf4j/rdf4j/issues/4446 , which marked the javax.servlet jstl dependency as "provided".

**Comment by adamretter** [comment]
I just tested, and I can add that this affects the War files of all `4.3.x` versions of rdf4j-server. I tested the following Docker images:

* 4.3.7 - Error when opening `localhost:8080/rdf4j-server`: `java.lang.ClassNotFoundException: javax.servlet.jsp.jstl.core.Config`
* 4.3.6 - Error when opening `localhost:8080/rdf4j-server`: `java.lang.ClassNotFoundException: javax.servlet.jsp.jstl.core.Config`
* 4.3.5 - Error when opening `localhost:8080/rdf4j-server`: `java.lang.ClassNotFoundException: javax.servlet.jsp.jstl.core.Config`
* 4.3.4 - Error when opening `localhost:8080/rdf4j-server`: `java.lang.ClassNotFoundException: javax.servlet.jsp.jstl.core.Config`
* 4.3.2 - Startup error in `catalina.log` - `Caused by: java.lang.NoClassDefFoundError: Lorg/apache/commons/logging/Log`
* 4.3.1 - No such docker image
* 4.3.0 - No such docker image

I don't think this is a docker related issue, as I also confirmed that `4.3.7` has the same problem outside of Docker on Tomcat versions `8.5.95` and `9.0.82`.

In summary, the latest working `rdf4j-server` appears to be: `4.2.3`.

**Comment by hmottestad** [comment]
Could be relevant:
 - https://stackoverflow.com/questions/68122661/cannot-get-jstl-tags-to-work-with-tomcat-9
 - https://stackoverflow.com/questions/10585608/implementing-jstl-1-2-1-in-a-servlet-3-0-container-on-tomcat-7
 - https://stackoverflow.com/questions/54299244/how-to-use-jstl-1-2-on-tomcat-8

I assume that the docker images don't add the jstl jar file. 

@adamretter It should be possible to add the jstl jar file yourself to your own tomcat. Have you tried that?

**Comment by erikgb** [comment]
> Adding the jstl jar file seems to have resolved the issue: #4814
> 
> @erikgb Could you give us some insight int0 why the jstl dependency in the pom.xml file is marked as provided? I know that tomcat with jakarta 10 should provide a tag lib implementation, but under a different uri. Since we are using the old sun uri then I'm not sure what we achieve by having the maven dependency marked as provided. I've only ever touched dedicated application servers while I was a student, and ever since then I've just used the integrated approach of spring-boot. Would the solution to the jstl issue be to have the maven dependency marked as compile? Or is it better to add the jstl jar file directly to the WEB-INF/lib directory as I've done in the PR above?

@hmottestad Thanks for the ping. The idea here is that all Java/Jakarta EE dependencies should be provided by the Java/Jakarta EE runtime. That's why I think it's correct to set the dependencies to scope=provided. That will avoid potential duplicate classes in the runtime classpath that could cause issues. But there might be some exceptions to this, especially in the "not commonly used" Java/Jakarta EE artifacts like JSTL. Anyway, I think it's correct to declare the dependencies as provided and add missing jar's to the application server runtime - if required to. Hope this clarifies my stand a bit.

**Comment by hmottestad** [comment]
Thanks for the quick reply. I'll add the jar file then, like in the PR.

**Comment by adamretter** [comment]
> @adamretter It should be possible to add the jstl jar file yourself to your own tomcat. Have you tried that?

I haven't tried that, and likely won't just yet as I switched down to using version `4.2.3`. If there is a new release of rdf4j that fixes the issue, I would be happy to test it though.

**Comment by hugovanversendaal** [comment]
I've added the jstl jar to my own image to test the solution by inserting the following line to my Dockerfile:

```
ADD https://repo1.maven.org/maven2/javax/servlet/jstl/1.2/jstl-1.2.jar ${CATALINA_HOME}/webapps/rdf4j-server/WEB-INF/lib
```

I can confirm that this solves the issue (using version `4.3.7` now).
